### ISSUE TITLE
PNG Promise does not appear to resolve (or reject)

### ISSUE THREAD

**Issue by alexpreynolds** [higlass/higlass#652]
_Commit_: 

HiGlass v1.5.3

#### Steps to reproduce
```
        let pngPromise = this.hgView.api.exportAsPngBlobPromise();
        console.log("pngPromise", pngPromise);
        pngPromise
          .then((blob) => {
            let reader = new FileReader(); 
            reader.addEventListener("loadend", function() {
              let array = new Uint8Array(reader.result.slice(0,8));
              /* do stuff with array */
            }); 
            reader.readAsArrayBuffer(blob);
          })
          .catch((err) => {
            throw Error(err);
          })
          .finally(function() {
            alert("PNG export attempt finished")
          });
```

#### Observed behavior
- The Promise `pngPromise` remains in `pending` mode and does not seem to resolve. Via console.log:

```
pngPromise 
  Promise {<pending>}
    __proto__: Promise
      [[PromiseStatus]]: "pending"
      [[PromiseValue]]: undefined
```

#### Expected behavior
- The Promise either resolves or rejects

**Comment by flekschas** [comment]
I am unable to reproduce this in `v1.5.7`. When I do:

```js
setTimeout(() => {
  hgApi.exportAsPngBlobPromise()
    .then((blob) => {
      const reader = new FileReader();
      reader.addEventListener('loadend', () => {
        const array = new Uint8Array(reader.result.slice(0, 8));
        console.log('all good', array);
      });
      reader.readAsArrayBuffer(blob);
    })
    .catch((err) => {
      throw Error(err);
    })
    .finally(() => {
      console.log('PNG export attempt finished');
    });
}, 2500);
```

I'll see:

```
HiGlass v1.5.7: http://higlass.io
PNG export attempt finished
all good Uint8Array(8) [137, 80, 78, 71, 13, 10, 26, 10]
```

Could you provide a complete example with the issue?

**Comment by alexpreynolds** [comment]
I upgraded HiGlass to v1.5.7 since filing this issue.

Here is the code that I have, relevant to PNG export (which fails) and SVG export (which works):

```
  onClickExportItemSelect = (name) => {
    switch (name) {
      ...
      case "png":
        console.log("this.hgView.api", this.hgView.api);
        let pngPromise = this.hgView.api.exportAsPngBlobPromise();
        console.log("pngPromise", pngPromise);
        pngPromise
          .then((blob) => {
            console.log(blob);
            let reader = new FileReader(); 
            reader.addEventListener("loadend", function() {
              let array = new Uint8Array(reader.result.slice(0,8));
              console.log(array); 
              console.log(new TextDecoder("iso-8859-2").decode(array));
              let pngBlob = new Blob(array, {type: "image/png"});
              saveAs(pngBlob, ["epilogos", params.genome, params.model, params.complexity, params.group, coord.chrLeft + '_' + coord.startLeft + '-' + coord.chrRight + '_' + coord.stopRight, "png"].join("."));
            }); 
            reader.readAsArrayBuffer(blob);
          })
          .catch(function(err) {
            throw Error(err);
          })
          .finally(function() {
            alert("PNG export attempt is complete");
          });
        break;
      case "svg":
        let svgStr = this.hgView.api.exportAsSvg();
        let coord = this.state.currentPosition;
        let params = this.state.hgViewParams;
        
        // cf. https://github.com/higlass/higlass/issues/651
        let fixedSvgStr = svgStr.replace('xmlns="http://www.w3.org/1999/xhtml"', '');
        
        let svgFile = new File([fixedSvgStr], ["epilogos", params.genome, params.model, params.complexity, params.group, coord.chrLeft + '_' + coord.startLeft + '-' + coord.chrRight + '_' + coord.stopRight, "svg"].join("."), {type: "image/svg+xml;charset=utf-8"});
        saveAs(svgFile);
        break;
      case "tabix":
        ...
        break;
      default:
        break;
    }
    this.setState({ exportIsVisible:false });
  }
```

SVG export works correctly, but the Promise in the PNG export case always stays in pending state. 

As you can see, I am using `this.hgView.api` in both cases to access the PNG and SVG export calls.

I also tried replacing my PNG export code with your `setTimeout()` code, and its `finally()` block never executes (I do not see the `PNG export attempt finished` message).

It is kind of difficult to set up a test site other than the development server I am working on at the moment. I could try to use the `higlass-docker` approach, but I don't think I can add the customized `higlass-multivec` plugin I am using. If it is useful to try out the development server, I will send you a link and credentials via Slack.

In the testing I am doing, my view only contains "stock" epilogos tracks. I am not testing export of the custom multivec track. In any case, this API call returns no errors or messages related to the execution of the export promise.

**Comment by flekschas** [comment]
Can you export to PNG via the UI?

**Comment by alexpreynolds** [comment]
Your PR fixes this issue for me: https://github.com/higlass/higlass/pull/663

**Comment by alexpreynolds** [comment]
Here is the fixed export code, for posterity:

```
  onClickExportItemSelect = (name) => {
    switch (name) {
      case "png":
        let pngPromise = this.hgView.api.exportAsPngBlobPromise();
        pngPromise
          .then((blob) => {
            console.log(blob);
            let reader = new FileReader(); 
            reader.addEventListener("loadend", function() {
              let array = new Uint8Array(reader.result);
              let pngBlob = new Blob([array], {type: "image/png"});
              saveAs(pngBlob, ["epilogos", params.genome, params.model, params.complexity, params.group, coord.chrLeft + '_' + coord.startLeft + '-' + coord.chrRight + '_' + coord.stopRight, "png"].join("."));
            }); 
            reader.readAsArrayBuffer(blob);
          })
          .catch(function(err) {
            throw Error(err);
          })
          .finally(function() {
            alert("PNG export attempt is complete");
          });
        break;
      /* ... */
      default:
        break;
    }
    this.setState({ exportIsVisible:false });
  }
```

**Comment by flekschas** [comment]
This might actually be related to the broken SVG export reported in #651 

Having fixed #651 in #713 the PNG export works and the promise gets resolved.

**Comment by flekschas** [comment]
This should be fixed in `v1.6`. If you still have issues please reopen.
